Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java

```diff
@@ -32,100 +30,76 @@
+                  Log.d(ContentValues.TAG, "Video chunk read unsuccessful with: " + e.getMessage());
+                }
+              }
+              return frames;
+            });
   }
 
-  public List<String> doOcr() {
-    AnalyseVideo ocrOnVideoTask = new AnalyseVideo();
-    try {
-      return ocrOnVideoTask.execute().get();
-    } catch (ExecutionException | InterruptedException e) {
-      e.printStackTrace();
-      return null;
-    }
-  }
-
-  //TODO: Replace AsyncTask with something more robust
-
-  /**
-   * Iterates on video frames issuing an Ocr request. Requests video-frame timestamp assignment.
-   * Calls LagCalculator upon Ocr completion of all video frame.
-   */
-  private class AnalyseVideo extends AsyncTask<Void, Void, List<String>> {
-
-    @Override
-    protected List<String> doInBackground(Void... values) {
-      TextRecognizer recognizer = TextRecognition.getClient();
-      List<String> resultsOCR = Arrays.asList(new String[frameList.size()]);
-      Collection<Task<Text>> ocrTasks = new ArrayList<>();
-      for (int i = 0; i < frameList.size(); i++) {
-        InputImage imageHolder = InputImage.fromBitmap(frameList.get(i), 0);
-        final int imageIndex = i;
-        ocrTasks.add(
-            recognizer
-                .process(imageHolder)
-                .addOnSuccessListener(
-                    visionText -> {
-                      resultsOCR.set(imageIndex, visionText.getText());
-                      Log.d(
-                          ContentValues.TAG,
-                          "Text detected at index:" + imageIndex + " " + visionText.getText());
-                    })
-                .addOnFailureListener(
-                    e -> Log.d(ContentValues.TAG, "Analyse failed with: " + e.getMessage())));
-      }
-      Task finalTask = Tasks.whenAllComplete(ocrTasks);
-      try {
-        Tasks.await(finalTask);
-        return resultsOCR;
-      } catch (InterruptedException | ExecutionException e) {
-        e.printStackTrace();
-        return null;
-      }
-    }
+  @RequiresApi(api = VERSION_CODES.N)
+  public CompletableFuture<ArrayList<Text>> doOcr() {
+    CompletableFuture<ArrayList<Text>> resultsOCR =
+        frameList.thenApply(
+            frames -> {
+              synchronized (frames) {
+                TextRecognizer recognizer = TextRecognition.getClient();
+                Collection<Task<Text>> ocrTasks = new ArrayList<>();
+                ArrayList<Text> ocrTexts = new ArrayList<>();
+                for (int i = 0; i < frames.size(); i++) {
+                  InputImage imageHolder = InputImage.fromBitmap(frames.get(i), 0);
+                  final int imageIndex = i;
+                  ocrTasks.add(
+                      recognizer
+                          .process(imageHolder)
+                          .addOnSuccessListener(
+                              visionText -> {
+                                ocrTexts.add(visionText);
```

Comment:
These callbacks may be executed out of order, you need to insert it in the correct location in the array rather than just appending.